### DISCUSSION TITLE
Sentence Generation Superfluous Whitespace

### DISCUSSION THREAD

**Discussion by MadcapJake** [mike-lischke/vscode-antlr4#217] [Q&A] [answered]
Why does my sentence generation cause tons of extra white space? Is there a way to clean it up?
```
0) te usen   te lika   ta o ten   tonsa        win ko     le na   te ukun   ta le tile   sala    win ka   wo tuko          . le peman     
1) o namuna    nenka ne   koton      . 
2) tu   te ne   latito    po kapu   inkatan       le mitin   eme     . 
```
https://git.sr.ht/~madcapjake/antlr-kokanu/tree/main/item/grammar/KokanuParser.g4

**Comment by mike-lischke** [comment, marked as answer]
Whitespaces are usually not quantitative elements, but more of fill elements, and appear in any cluster size. The sentence generation, however, generates on a single space (not tab, no linebreak etc.) between every generated element. Because of the recursive and iterative nature of parser grammars, elements might not have been created actually, but their surrounding spaces, which creates a sequence of  them. 

In any case, these (extra) spaces have no impact on the parsability of the generated text. It's still valid input for the given grammar. I recommend adding a post processing job to get rid of elements you don't want.
